Build fails after updating to 2.1.x if there is the nullish coalescing operator (??) in a SFC template (Vue)

<!--
中文用户请注意：请仔细阅读以下模版，如果不遵循模版，issue 将会被直接关闭。
-->

**⚠️ IMPORTANT ⚠️ Please do not ignore this template. If you do, your issue will be closed immediately.**

- [x] Read [the docs](https://vitejs.dev/guide/).
- [x] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/vue-next instead.
- [x] This is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).

## Describe the bug

After upgrading Vite to version `2.1.1`, my builds started failing during the transform phase with an error message implying that Vite could not make my code compatible with the targeted browser versions.

I found out that this was at least caused by the use of `??` in my SFC templates, which was not a surprise because the dev version of my app never worked in Chrome since this operator is not supported. This was not too problematic, since I use Firefox during developpment and I always could test my app in Chrome after the build phase thanks to the `serve` command, but now that the build is failing it is much more bothersome.

Downgrading to Vite `2.0.5` solves the problem, and using `??` in the SFC `<script>` works just fine no matter the Vite version.

## Reproduction

https://github.com/IonianPlayboy/vite-build-failure-repro

The repo was generated with `npm init @vitejs/app@latest`, choosing the `vue-ts` template. The only change was adding the nullish coalescing operator in `HelloWorld.vue` :
```html
<h1>{{ msg ?? "" }}</h1>
```

The `master` branch has Vite at version `2.1.1`, and the build command is failing. The ` vite@2.0.5` branch has Vite at version `2.0.5`, and the build command works just fine.

## System Info

- `vite` version: 2.1.1
- Operating System: Windows 10
- Node version: 14.15.4
- Package manager (npm/yarn/pnpm) and version: Yarn 1.22.5

## Logs (Optional if provided reproduction)

```sh
yarn build --debug
yarn run v1.22.5
warning package.json: No license field
$ vue-tsc --noEmit && vite build --debug
  vite:config bundled config file loaded in 358ms +0ms
  vite:config using resolved config: {
  vite:config   plugins: [
  vite:config     'alias',
  vite:config     'vite:dynamic-import-polyfill',
  vite:config     'vite:resolve',
  vite:config     'vite:html',
  vite:config     'vite:css',
  vite:config     'vite:esbuild',
  vite:config     'vite:json',
  vite:config     'vite:wasm',
  vite:config     'vite:worker',
  vite:config     'vite:asset',
  vite:config     'vite:vue',
  vite:config     'vite:define',
  vite:config     'vite:css-post',
  vite:config     'vite:build-html',
  vite:config     'commonjs',
  vite:config     'vite:data-uri',
  vite:config     'rollup-plugin-dynamic-import-variables',
  vite:config     'vite:import-analysis',
  vite:config     'vite:esbuild-transpile',
  vite:config     'vite:terser',
  vite:config     'vite:reporter'
  vite:config   ],
  vite:config   build: {
  vite:config     target: [ 'es2019', 'edge16', 'firefox60', 'chrome61', 'safari11' ],
  vite:config     polyfillDynamicImport: true,
  vite:config     outDir: 'dist',
  vite:config     assetsDir: 'assets',
  vite:config     assetsInlineLimit: 4096,
  vite:config     cssCodeSplit: true,
  vite:config     sourcemap: false,
  vite:config     rollupOptions: {},
  vite:config     commonjsOptions: { include: [Array], extensions: [Array] },
  vite:config     minify: 'terser',
  vite:config     terserOptions: {},
  vite:config     cleanCssOptions: {},
  vite:config     write: true,
  vite:config     emptyOutDir: null,
  vite:config     manifest: false,
  vite:config     lib: false,
  vite:config     ssr: false,
  vite:config     ssrManifest: false,
  vite:config     brotliSize: true,
  vite:config     chunkSizeWarningLimit: 500
  vite:config   },
  vite:config   define: { __VUE_OPTIONS_API__: true, __VUE_PROD_DEVTOOLS__: false },
  vite:config   ssr: { external: [ 'vue', '@vue/server-renderer' ] },
  vite:config   configFile: 'C:/Users/Alex Lafeuille/Documents/vite-build-failure-repro/vite.config.ts',
  vite:config   inlineConfig: {
  vite:config     root: undefined,
  vite:config     base: undefined,
  vite:config     mode: undefined,
  vite:config     configFile: undefined,
  vite:config     logLevel: undefined,
  vite:config     clearScreen: undefined,
  vite:config     build: {}
  vite:config   },
  vite:config   root: 'C:/Users/Alex Lafeuille/Documents/vite-build-failure-repro',
  vite:config   base: '/',
  vite:config   resolve: { dedupe: undefined, alias: [ [Object] ] },
  vite:config   publicDir: 'C:\\Users\\Alex Lafeuille\\Documents\\vite-build-failure-repro\\public',
  vite:config   command: 'build',
  vite:config   mode: 'production',
  vite:config   isProduction: true,
  vite:config   optimizeCacheDir: 'C:\\Users\\Alex Lafeuille\\Documents\\vite-build-failure-repro\\node_modules\\.vite',   
  vite:config   server: {},
  vite:config   env: { BASE_URL: '/', MODE: 'production', DEV: false, PROD: true },
  vite:config   assetsInclude: [Function: assetsInclude],
  vite:config   logger: {
  vite:config     hasWarned: false,
  vite:config     info: [Function: info],
  vite:config     warn: [Function: warn],
  vite:config     error: [Function: error],
  vite:config     clearScreen: [Function: clearScreen]
  vite:config   },
  vite:config   createResolver: [Function: createResolver]
  vite:config } +14ms
vite v2.1.1 building for production...
✓ 15 modules transformed.
  vite:esbuild esbuild error with options used:  {
  loader: 'js',
  sourcemap: true,
  sourcefile: 'assets/vendor.6801b2a0.js',
  target: [ 'es2019', 'edge16', 'firefox60', 'chrome61', 'safari11' ],
  minify: false
} +0ms
error during build:
Error: Transform failed with 64 errors:
assets/vendor.6801b2a0.js:103:70: error: Transforming destructuring to the configured target environment ("chrome61", "edge16", "es2019", "firefox60", "safari11") is not supported yet
assets/vendor.6801b2a0.js:269:10: error: Transforming destructuring to the configured target environment ("chrome61", "edge16", "es2019", "firefox60", "safari11") is not supported yet
assets/vendor.6801b2a0.js:548:10: error: Transforming destructuring to the configured target environment ("chrome61", "edge16", "es2019", "firefox60", "safari11") is not supported yet
assets/vendor.6801b2a0.js:588:10: error: Transforming destructuring to the configured target environment ("chrome61", "edge16", "es2019", "firefox60", "safari11") is not supported yet
assets/vendor.6801b2a0.js:606:10: error: Transforming destructuring to the configured target environment ("chrome61", "edge16", "es2019", "firefox60", "safari11") is not supported yet
...
    at failureErrorWithLog (C:\Users\Alex Lafeuille\Documents\vite-build-failure-repro\node_modules\esbuild\lib\main.js:1195:15)
    at C:\Users\Alex Lafeuille\Documents\vite-build-failure-repro\node_modules\esbuild\lib\main.js:1056:33
    at C:\Users\Alex Lafeuille\Documents\vite-build-failure-repro\node_modules\esbuild\lib\main.js:566:9
    at handleIncomingPacket (C:\Users\Alex Lafeuille\Documents\vite-build-failure-repro\node_modules\esbuild\lib\main.js:655:9)
    at Socket.readFromStdout (C:\Users\Alex Lafeuille\Documents\vite-build-failure-repro\node_modules\esbuild\lib\main.js:533:7)
    at Socket.emit (events.js:315:20)
    at addChunk (internal/streams/readable.js:309:12)
    at readableAddChunk (internal/streams/readable.js:284:9)
    at Socket.Readable.push (internal/streams/readable.js:223:10)
    at Pipe.onStreamRead (internal/stream_base_commons.js:188:23)
error Command failed with exit code 1.
```

--- a/packages/vite/package.json
+++ b/packages/vite/package.json
@@ -44,7 +44,7 @@
   },
   "//": "READ .github/contributing.md to understand what to put under deps vs. devDeps!",
   "dependencies": {
-    "esbuild": "^0.9.2",
+    "esbuild": "^0.9.3",
     "postcss": "^8.2.1",
     "resolve": "^1.19.0",
     "rollup": "^2.38.5"

--- a/packages/vite/src/node/build.ts
+++ b/packages/vite/src/node/build.ts
@@ -447,7 +447,7 @@ async function doBuild(
       }
       return res
     } else {
-      return generate(outputs)
+      return await generate(outputs)
     }
   } catch (e) {
     config.logger.error(

--- a/packages/vite/src/node/build.ts
+++ b/packages/vite/src/node/build.ts
@@ -220,7 +220,9 @@ export function resolveBuildOptions(raw?: BuildOptions): ResolvedBuildOptions {
   // handle special build targets
   if (resolved.target === 'modules') {
     // https://caniuse.com/es6-module
-    resolved.target = ['es2019', 'edge16', 'firefox60', 'chrome61', 'safari11']
+    // edge18 according to js-table (destructuring is not supported in edge16)
+    // https://github.com/evanw/esbuild/blob/d943e89e50696647d6c89ae623ddfdf564ad3cfc/internal/compat/js_table.go#L84
+    resolved.target = ['es2019', 'edge18', 'firefox60', 'chrome61', 'safari11']
   } else if (resolved.target === 'esnext' && resolved.minify !== 'esbuild') {
     // esnext + terser: limit to es2019 so it can be minified by terser
     resolved.target = 'es2019'

